### DISCUSSION TITLE
Dynamically generating nested queries

### DISCUSSION THREAD

**Discussion by MarkNicholls** [linq2db/linq2db#3490] [Q&A]
Posted from stack overflow because the moderator closed the question just before 
Svyatoslav Danyliv was about to reply - he suggested posting here, so he could respond...it IS a question based on EF core - apologies...I may well port to linq2db as the startup up on EF core is horrifically slow.

The ultimate objective is to have a piece of code that you can feed definitions of simple queries to and it generate data.

so given i can create this query
```cs
var query =
    from cha in ds.Channels
    select new object[] 
    { 
        cha.ChanLanguagestreamlocking, 
        cha.ChanMinimumtrailerduration,
        from link in cha.Channelaudiolanguagelinks
        select new object[]
        {
            link.ObjDatecreated
        }
    };

var data = query.ToArray();
```
how would i create it dynamically?

well i can get this far using a class stolen from another stack overflow question
```cs
public class SelectList<TSource>
{
    private List<LambdaExpression> members = new List<LambdaExpression>();
    public SelectList<TSource> Add<TValue>(Expression<Func<TSource, TValue>> selector)
    {
        members.Add(selector);
        return this;
    }

    public Expression<Func<TSource, object[]>> ToDynamicColumns()
    {
        var parameter = Expression.Parameter(typeof(TSource), "e");
        return Expression.Lambda<Func<TSource, object[]>>(
            Expression.NewArrayInit(
                typeof(object),
                members.Select(m =>
                    Expression.Convert(Expression.Invoke(m, parameter), typeof(object))
                )
            ),
            parameter);
    }
}
```
Imagine a case where the definition of which 'columns' to return is passed as a tree at runtime, this can then be mapped to predefined typessafe lambdas that define which columns to return and then the expression is created at runtime. So a hard coded version of this technique is this:
```cs
var channelColumns = new SelectList<Channel>();
channelColumns.Add(c => c.ChanLanguagestreamlocking);
channelColumns.Add(c => c.ChanMinimumtrailerduration);

var channelQuery = 
    ds.Channels.Select(channelColumns.ToDynamicColumns());
var bar = query.ToArray();
```
i.e. i can generate dynamic queries from the 'root' concept, but how do i generate the nested data.

If i do the obvious i.e. this
```cs
var audioColumns = new SelectList<Channelaudiolanguagelink>();
audioColumns.Add(a => a.ObjDatecreated);
var channelColumns = new SelectList<Channel>();
channelColumns.Add(c => c.ChanLanguagestreamlocking);
channelColumns.Add(c => c.ChanMinimumtrailerduration);
// next line causes an error
// Error CS1929  'ICollection<Channelaudiolanguagelink>' does not contain a definition for
// 'Select' and the best extension method overload
// 'Queryable.Select<Channel, object[]>(IQueryable<Channel>, Expression<Func<Channel, object[]>>)' requires a receiver of type 'IQueryable<Channel>' CSharpDb2Raw    C:\Users\mark.nicholls\source\repos\scaffold2\CSharpDb2Raw\Program.cs   57  Active
channelColumns.Add(c => c.Channelaudiolanguagelinks.Select(channelColumns.ToDynamicColumns()));
var channelQuery = 
    ds.Channels.Select(channelColumns.ToDynamicColumns());
var bar = query.ToArray();
```
the error makes perfect sense.

c.Channelaudiolanguagelinks is an ICollection<T> and so the select is looking for a Func<T,U> and I've given it an Expression<Func<T,U>>

(I don't really understand Expressions!)

**Comment by sdanyliv** [comment]
Well, easiest way:
```cs
channelColumns.Add(c => c.Channelaudiolanguagelinks.AsQueryable().Select(channelColumns.ToDynamicColumns()));
```
But for sure we can invent much more usable extensions. We can transform `ExpressionTree` to anything.

**Reply by MarkNicholls** [reply]
hmmm that makes sense, but im getting type errors....wait a sec

**Reply by MarkNicholls** [reply]
'Queryable.Select<TSource, TResult>(IQueryable<TSource>, Expression<Func<TSource, int, TResult>>)' cannot be inferred from the usage. Try specifying the type arguments explicitly.

**Reply by sdanyliv** [reply]
I have tested on similar query without problem. Check another approach.

**Reply by MarkNicholls** [reply]
channelColumns.Add(c =>
                {
                    IQueryable<Channelaudiolanguagelink>? q = c.Channelaudiolanguagelinks.AsQueryable();
                    Expression<Func<Channel, object[]>>? columns = channelColumns.ToDynamicColumns();
                    return q.Select(columns);
                });

**Reply by MarkNicholls** [reply]
these are my types

**Reply by sdanyliv** [reply]
You cannot do that in ExpressionTree!
```cs
channelColumns.Add(c =>c.Channelaudiolanguagelinks.AsQueryable().Select(channelColumns.ToDynamicColumns()));
```
Also check that you have used bad mapper, you have to use Audio columns SelectList

**Reply by MarkNicholls** [reply]
no wait theres a typo

**Reply by MarkNicholls** [reply]
channelColumns.Add(c => c.Channelaudiolanguagelinks.AsQueryable().Select(audioColumns.ToDynamicColumns()));

**Reply by sdanyliv** [reply]
This one should work.

**Reply by MarkNicholls** [reply]
works a treat!
thanks very very much, I basically don't understand expressions at all, there's some sort of compiler magic going on.
Your other answer does look more complicated! does it have some inherent advantage?

**Reply by MarkNicholls** [reply]
do u want to post it on stack overflow? or shall I....I must admit to being a bit irritated with the moderator, seems to be an growing theme in some groups, they want to write the question and the answer.

**Reply by sdanyliv** [reply]
> does it have some inherent advantage?

Just if you do not want to use `AsQueryable()`.

 Anyway issue is reopened. And I can post simple answer. If you explain what you are trying to achieve at the end maybe I can give you best solution, I know `ExpressionTree` very well ;)

**Reply by MarkNicholls** [reply]
well, I am using as system that defines queries as trees of Select lists, where each relationship is mapped to a nested value, all mapped to XML, probably like a very simpe version of GraphQl
This system is incredibly slow, because the root query is a single query but the children are all retrieved individually. literally thousands of SQL statements to retrieve thousands of records. I was hoping to build something very similar but a sinlge SQL statement was generated and executed, without resorting to SQL/Dapper, all in LINQ. I can probably auto generate the Schema as a tree, adn then people would simply select the fields they want to retrieve pass it to the data retriever, it maps it to a single SQL and executes it and then serialise the output to XML or JSON. probably be about 1000x quicker than the existing implmentation.

**Reply by sdanyliv** [reply]
Sorry, missed message. Yes, why not. I have consulted something similar 6 years ago. And linq2db is suitable for such queries instead of EF. Maybe you will show XML?

**Reply by MarkNicholls** [reply]
well, ive done a basic mock up based on your helpful answer and EF. You can drive a query simply with a tree, which is more easily expressed as a set of paths and a root table so (in psuedo F#)
{
RootTable = "Table1"
Paths = [ 
 "attribute1"
 "attribute2"
 "referencetoTable2/attribute3"
 "referencetoTable2/attribute4"
]
}

you process that into a tree and a root concept, and then write something that to create the selectlists recusively from the tree using the code you've given me, and you get a nice hierachical object[] back where some objects in the array can be object[]
I've got that much working.
I'll probably make it capable of processing that into JSON/XML with elements that correspond to the table names.
I need to work out how to include a where clause on the root table.
I'm slightly concerned theres a point where the resulting SQL can be too complex, I did a very primitive version of this about 15 years ago, by constructing SQL strings, and eventually SQL server complained that the SQL was too complex.

**Comment by sdanyliv** [comment]
Another solution. We can add additional method, which returns `Func<TSource, object[]>`. For example `Func<TSource, object[]> ToDynamic()`.

Implementation may looks like complicated but it's not ;)
```cs
channelColumns.Add(c => c.Channelaudiolanguagelinks.Select(audioColumns.ToDynamic()));
```
```cs
public class SelectList<TSource>
{
    private List<LambdaExpression> members = new();
    public SelectList<TSource> Add<TValue>(Expression<Func<TSource, TValue>> selector)
    {
        members.Add(selector);
        return this;
    }

    public Expression<Func<TSource, object[]>> ToDynamicColumns()
    {
        var parameter = Expression.Parameter(typeof(TSource), "e");
        Expression body = 
            Expression.NewArrayInit(
                typeof(object),
                members.Select(m =>
                    Expression.Convert(Expression.Invoke(m, parameter), typeof(object))
                )
            );

        body = new ExpressionExpander().Visit(body);

        var lambda = Expression.Lambda<Func<TSource, object[]>>(body, parameter);
        return lambda;
    }

    public Func<TSource, object[]> ToDynamic()
    {
        throw new NotImplementedException();
    }

    class ExpressionExpander : ExpressionVisitor
    {
        protected override Expression VisitMethodCall(MethodCallExpression node)
        {
            var method = node.Method;
            if (method.DeclaringType != null && method.DeclaringType.IsGenericType &&
                method.DeclaringType.GetGenericTypeDefinition() == typeof(SelectList<>) && method.Name == nameof(SelectList<object>.ToDynamic))
            {
                var callMethod = Expression.Call(node.Object,
                    method.DeclaringType.GetMethod(nameof(SelectList<object>.ToDynamicColumns)));

                var generatedLambda = (LambdaExpression)callMethod.EvaluateExpression();

                return Visit(generatedLambda);
            }

            return base.VisitMethodCall(node);
        }
    }
}
```
And needed helper extension:
```cs
public static class ExpressionExtensions
{
    public static object? EvaluateExpression(this Expression? expr)
    {
        if (expr == null)
            return null;

        switch (expr.NodeType)
        {
            case ExpressionType.Default :
                return !expr.Type.IsNullableType() ? Activator.CreateInstance(expr.Type) : null;

            case ExpressionType.Constant:
                return ((ConstantExpression)expr).Value;

            case ExpressionType.Convert:
            case ExpressionType.ConvertChecked:
            {
                var unary = (UnaryExpression)expr;
                var operand = unary.Operand.EvaluateExpression();
                if (operand == null)
                    return null;
                break;
            }

            case ExpressionType.MemberAccess:
            {
                var member = (MemberExpression) expr;

                if (member.Member.IsFieldEx())
                    return ((FieldInfo)member.Member).GetValue(member.Expression.EvaluateExpression());

                if (member.Member is PropertyInfo propertyInfo)
                {
                    var obj = member.Expression.EvaluateExpression();
                    if (obj == null)
                    {
                        if (propertyInfo.IsNullableValueMember())
                            return null;
                        if (propertyInfo.IsNullableHasValueMember())
                            return false;
                    }
                    return propertyInfo.GetValue(obj, null);
                }
                    
                break;
            }
            case ExpressionType.Call:
            {
                var mc = (MethodCallExpression)expr;
                var arguments = mc.Arguments.Select(EvaluateExpression).ToArray();
                var instance  = mc.Object.EvaluateExpression();

                if (instance == null && mc.Method.IsNullableGetValueOrDefault())
                    return null;
                    
                return mc.Method.Invoke(instance, arguments);
            }
        }

        var value = Expression.Lambda(expr).Compile().DynamicInvoke();
        return value;
    }
}
```

**Reply by MarkNicholls** [reply]
I'm not clear as to how to use this code from the client side

**Reply by MarkNicholls** [reply]
I've got to go an do some 'proper' work now...but i'll take a look in an hour or two
